pkg/metrics/metrics.go

```diff
@@ -565,27 +576,31 @@ func (c *Counter) Query(query string) (*AlertResponse, error) {
 	return result, nil
 }
 
-func prometheusToken(c client.Client) (*string, error) {
-	sl := &corev1.SecretList{}
-	err := c.List(context.TODO(), sl, client.InNamespace(MonitoringNS))
+func NewKubernetesClient() (*kubernetes.Clientset, error) {
+	config, err := rest.InClusterConfig()
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("unable to get config due to: %w", err)
 	}
 
-	token := ""
-	for _, secret := range sl.Items {
-		if strings.HasPrefix(secret.Name, "prometheus-k8s-token") {
-			tokendata := secret.Data[corev1.ServiceAccountTokenKey]
-			token = string(tokendata)
-			break
-		}
+	k8sClient, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("unable to creat kubernets clientSet to: %w", err)
 	}
 
-	if len(token) == 0 {
-		return nil, fmt.Errorf("failed to find token secret for prometheus-k8s SA")
-	}
+	return k8sClient, nil
+}
 
-	return &token, nil
+// RequestPrometheusServiceAccountAPIToken returns a time-bound (1hr) API token for the prometheus service account.
+func RequestPrometheusServiceAccountAPIToken(kc kubernetes.Clientset) (string, error) {
+	expirationSeconds := int64(1 * time.Hour / time.Second)
+	req, err := kc.CoreV1().ServiceAccounts(MonitoringNS).CreateToken(context.TODO(), promApp,
```

Comment:
As discussed, we should have the legacy mechanism of referring the prometheus token too if it is present. So that will keep the mechanism for clusters < 4.16 intact and 4.16 onwards it will request for new short lived token as required.

Reply:
Updated